### PR TITLE
[#IOPID-867] magic link on login email

### PR DESCRIPTION
**Pull request by arcogabbo**
This PR depends on https://github.com/pagopa/io-web-profile-backend/pull/25 and https://github.com/pagopa/io-infra/pull/714

#### List of Changes

**NOTE**: a new required env variable has been introduced: `MAGIC_LINK_SERVICE_API_KEY`.
* [[#IOPID-867] introduced new login template and renamed previous one](https://github.com/pagopa/io-functions-app/commit/1a28cc6a1bf7c4e71eb0611a0442d3442d1f9a86)
* [[#IOPID-867] introduced new template in activity logic](https://github.com/pagopa/io-functions-app/commit/814867b49a58a891dc9f9d474c0bb55408733a7b)
* [[#IOPID-867] implemented actual client](https://github.com/pagopa/io-functions-app/commit/c3aae3eb8c44ebeef05bd22703b02397a82f5e1f)
* [[#IOPID-867] various enhancements](https://github.com/pagopa/io-functions-app/commit/b80deac356f820ecf23d6d987e42371e8562ab99)

#### Motivation and Context

introduce a new login email whenever we have a magic_link

#### How Has This Been Tested?

unit tests, integration tests via io-mock project

#### Screenshots (if appropriate):
fallback email(when we can not obtain the magic link from the service):
Pasted image 20231012143616

new login email template with magic link:
Pasted image 20231012144438

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: SendTemplatedLoginEmailActivity/__tests__/handler.test.ts [hunk 1]
```diff
@@ -36,18 +45,27 @@ describe("SendTemplatedLoginEmailActivity", () => {
...
-    const result = await handler(context as any, aValidPayload);
+    const templateFunction = jest.spyOn(
+      payload.magic_link ? mailTemplate : fallbackMailTemplate,
+      "apply"
+    );
+
+    const result = await handler(context as any, payload);
 
     expect(result.kind).toEqual("SUCCESS");
+    expect(templateFunction).toHaveBeenCalledTimes(1);
```
Comment (gquadrati) [inline SendTemplatedLoginEmailActivity/__tests__/handler.test.ts:68]: I'd rather check both calls like (spying both `templateFunction` and `fallbackTemplateFunction`)

```typescript
    expect(templateFunction).toHaveBeenCalledTimes(payload.magic_link ? 1 : 0);
    expect(fallbackTemplateFunction).toHaveBeenCalledTimes(
      payload.magic_link ? 0 : 1
    );
 ```

File: GetMagicCodeActivity/__tests__/handler.test.ts [hunk 2]
```diff
@@ -1,30 +1,37 @@
...
     );
-    expect(TransientNotImplementedFailure.is(result)).toEqual(true);
+    expect(ActivityResultSuccess.is(result)).toEqual(true);
   });
 
   it("should return a FAILURE when the input is not valid", async () => {
-    const result = await getActivityHandler(mockMagicCodeServiceClient)(
+    const result = await getActivityHandler(mockMagicLinkServiceClient)(
       context as any,
       {}
```
Comment (gquadrati) [inline GetMagicCodeActivity/__tests__/handler.test.ts:36]: Include test cases for other failure scenarios

**Review by gquadrati — APPROVED** [review]
LGTM, I've just posted two minor comments

File: GetMagicCodeActivity/__tests__/handler.test.ts (hunk 2 above)
Comment (arcogabbo) [reply GetMagicCodeActivity/__tests__/handler.test.ts]: added them in https://github.com/pagopa/io-functions-app/pull/285/commits/24e1bde8503879223b3d07a822bf1e8638544508

**Pull request merged**